### PR TITLE
Add capability to enable vulnerability assessments for Azure Synapse SQL Pools

### PR DESCRIPTION
**Pull request by owenfarrell**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

### Summary

This pull request adds support for managing vulnerability assessments for Synapse workspaces and Synapse workspace SQL pools. The implementation and the acceptance tests are largely derived from analogous implementations for SQL servers and SQL databases.

Vulnerability assessments are dependent on the definition of security alert policies, so this pull request includes the definition of **_those_** resources as well.

This PR likely conflicts in some common client/registration code with #12952, #13044, and/or #13179.

Output from acceptance testing:

```
$ make acctests SERVICE='synapse' TESTARGS='-run="^TestAccSynapse(SqlPool|Workspace)(SecurityAlertPolicy|VulnerabilityAssessment)_"' TESTTIMEOUT='120m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/synapse -run="^TestAccSynapse(SqlPool|Workspace)(SecurityAlertPolicy|VulnerabilityAssessment)_" -timeout 120m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccSynapseSqlPoolSecurityAlertPolicy_basic
=== PAUSE TestAccSynapseSqlPoolSecurityAlertPolicy_basic
=== RUN   TestAccSynapseSqlPoolSecurityAlertPolicy_update
=== PAUSE TestAccSynapseSqlPoolSecurityAlertPolicy_update
=== RUN   TestAccSynapseSqlPoolVulnerabilityAssessment_basic
=== PAUSE TestAccSynapseSqlPoolVulnerabilityAssessment_basic
=== RUN   TestAccSynapseSqlPoolVulnerabilityAssessment_update
=== PAUSE TestAccSynapseSqlPoolVulnerabilityAssessment_update
=== RUN   TestAccSynapseWorkspaceSecurityAlertPolicy_basic
=== PAUSE TestAccSynapseWorkspaceSecurityAlertPolicy_basic
=== RUN   TestAccSynapseWorkspaceSecurityAlertPolicy_update
=== PAUSE TestAccSynapseWorkspaceSecurityAlertPolicy_update
=== RUN   TestAccSynapseWorkspaceVulnerabilityAssessment_basic
=== PAUSE TestAccSynapseWorkspaceVulnerabilityAssessment_basic
=== RUN   TestAccSynapseWorkspaceVulnerabilityAssessment_update
=== PAUSE TestAccSynapseWorkspaceVulnerabilityAssessment_update
=== CONT  TestAccSynapseSqlPoolSecurityAlertPolicy_basic
=== CONT  TestAccSynapseWorkspaceVulnerabilityAssessment_update
=== CONT  TestAccSynapseWorkspaceSecurityAlertPolicy_basic
=== CONT  TestAccSynapseWorkspaceVulnerabilityAssessment_basic
=== CONT  TestAccSynapseWorkspaceSecurityAlertPolicy_update
=== CONT  TestAccSynapseSqlPoolSecurityAlertPolicy_update
=== CONT  TestAccSynapseSqlPoolVulnerabilityAssessment_update
=== CONT  TestAccSynapseSqlPoolVulnerabilityAssessment_basic
--- PASS: TestAccSynapseWorkspaceVulnerabilityAssessment_basic (781.00s)
--- PASS: TestAccSynapseWorkspaceSecurityAlertPolicy_basic (849.34s)
--- PASS: TestAccSynapseWorkspaceVulnerabilityAssessment_update (871.95s)
--- PASS: TestAccSynapseSqlPoolVulnerabilityAssessment_update (898.32s)
--- PASS: TestAccSynapseSqlPoolSecurityAlertPolicy_update (912.97s)
--- PASS: TestAccSynapseWorkspaceSecurityAlertPolicy_update (921.62s)
--- PASS: TestAccSynapseSqlPoolVulnerabilityAssessment_basic (1011.34s)
--- PASS: TestAccSynapseSqlPoolSecurityAlertPolicy_basic (1014.87s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/synapse       1017.121s
```

### CODE REVIEW

File: internal/services/synapse/synapse_sql_pool_security_alert_policy_resource.go [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+			},
+
+			"retention_days": {
+				Type:         pluginsdk.TypeInt,
+				Optional:     true,
+				Default:      0,
+				ValidateFunc: validation.IntAtLeast(0),
+			},
+
+			"state": {
```
Comment (stephybun) [inline internal/services/synapse/synapse_sql_pool_security_alert_policy_resource.go:83]: Could we be explicit here and rename this to 
```suggestion
			"policy_state": {
```

File: internal/services/synapse/synapse_workspace_security_alert_policy_resource.go [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+			},
+
+			"retention_days": {
+				Type:         pluginsdk.TypeInt,
+				Optional:     true,
+				Default:      0,
+				ValidateFunc: validation.IntAtLeast(0),
+			},
+
+			"state": {
```
Comment (stephybun) [inline internal/services/synapse/synapse_workspace_security_alert_policy_resource.go:83]: Same here 
```suggestion
			"policy_state": {
```

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks for this PR @owenfarrell. For the most part this looks good, just a few suggestions regarding the property names in the schema. Would also be good if the merge conflict could be resolved 🙂.

**Comment by owenfarrell** [comment]
@stephybun - Okay, the schema changes ended up being a bit meatier than I expected. But I think I've got all the code and documentation updates done. I've rerun the acceptance tests and everything still passes. And I've updated all of the error messages to align with the pattern change you requested.

Let me know if there's anything else!

**Comment by stephybun** [comment]
Thanks @owenfarrell, LGTM 🚀

**Review by stephybun — APPROVED** [review]

**Pull request merged**